File: kotlin-examples/src/main/java/javaVariant/2ConditionsAndTypeSwitch.java
package javaVariant;

import java.sql.SQLException;
import java.util.Locale;

class Conditions {
    public Locale getDefaultLocale(String deliveryArea){
        if (deliveryArea.equals("germany") || deliveryArea.equals("austria") || deliveryArea.equals("switzerland")) {
            return Locale.GERMAN;
        }
        if (deliveryArea.equals("usa") || deliveryArea.equals("great britain") || deliveryArea.equals("australia")) {
            return Locale.ENGLISH;
        }
        throw new IllegalArgumentException("Unsupported deliveryArea " + deliveryArea);
    }
    //or via switch and fall-through


    //verbose. annoying type cast.
    public String getExceptionMessage(Exception exception){
        if (exception instanceof MyLabeledException){
            return ((MyLabeledException) exception).getLabel();
        } else if (exception instanceof SQLException){
            return exception.getMessage() + ". state: " + ((SQLException) exception).getSQLState();
        } else {
            return exception.getMessage();
        }
    }
}

class MyLabeledException extends RuntimeException{
    private String label;

    public String getLabel() {
        return label;
    }

    public void setLabel(String label) {
        this.label = label;
    }
}

File: kotlin-examples/src/main/java/kotlinVariant/2ConditionsAndTypeSwitch.kt
package kotlinVariant

import java.sql.SQLException
import java.util.Locale

//"when" is extremely powerful construct is Kotlin. It's much more than just a switch.

fun getDefaultLocale(deliveryArea: String): Locale {
    when (deliveryArea){
        "germany", "austria", "switzerland" -> return Locale.GERMAN
        "usa", "great britain", "australia" -> return Locale.ENGLISH
        else -> throw IllegalArgumentException("Unsupported deliveryArea $deliveryArea") //string interpolation
    }
}
// or even shorter as a single expression function:
fun getDefaultLocale2(deliveryArea: String) = when (deliveryArea){
    "germany", "austria", "switzerland" -> Locale.GERMAN
    "usa", "great britain", "australia" -> Locale.ENGLISH
    else -> throw IllegalArgumentException("Unsupported deliveryArea $deliveryArea")
}

fun getExceptionMessage(exception: Exception) = when (exception){
    //concise type switches
    is MyLabeledException -> exception.label //smart cast to MyLabeledException -> we can call label directly.
    is SQLException -> "${exception.message}. state: ${exception.sqlState}" //string interpolation
    else -> exception.message
}
class MyLabeledException(val label: String) : RuntimeException(label)